fem/gridfunc.cpp

```diff
@@ -2410,30 +2447,73 @@ void GridFunction::ProjectCoefficient(VectorCoefficient &vcoeff)
    }
    else
    {
-      // Define and assemble linear form
-      LinearForm b(fes);
-      b.AddDomainIntegrator(new VectorFEDomainLFIntegrator(vcoeff));
-      b.Assemble();
+      constexpr real_t signal = std::numeric_limits<real_t>::min();
+      for (int i = 0; i < fes->GetNE(); i++)
+      {
+         fes->GetElementVDofs(i, vdofs, doftrans);
+         vals.SetSize(vdofs.Size());
+         vals = signal;
+         fes->GetFE(i)->Project(vcoeff, *fes->GetElementTransformation(i), vals);
+         doftrans.TransformPrimal(vals);
+         // Remove undefined dofs
+         // The knot location (either Botella, Demko or Greville point)
+         // where the NURBS dof are evaluated might fall outside of the
+         // domain of the element. In that case the value is not set, and
+         // the value remains the signal value.
+         int s = 0;
+         for (int ii = 0; ii < vals.Size(); ii++)
+         {
+            if (vals[ii] != signal)
+            {
+               vdofs[s] = vdofs[ii];
+               vals(s) = vals(ii);
+               s++;
+            }
+         }
+         vdofs.SetSize(s);
+         vals.SetSize(s);
+
+         // Add reduced dofs to global vector
+         SetSubVector(vdofs, vals);
+      }
+   }
+}
+
+void GridFunction::ProjectCoefficientGlobalL2(VectorCoefficient &vcoeff,
+                                              real_t rtol, int iter)
+{
+   // Define and assemble linear form
+   LinearForm b(fes);
+   BilinearForm a(fes);
 
-      // Define and assemble bilinear form
-      BilinearForm a(fes);
+   // Dimension argument to GetRangeType is arbitrary to be 3, could also be 2.
+   if ((fes->FEColl()->GetRangeType(3) == mfem::FiniteElement::VECTOR) ||
+       (fes->FEColl()->GetRangeType(2) == mfem::FiniteElement::VECTOR))
```

Comment:
This check should not be randomly using 3 or 2 ... it should use precisely the dimension of the mesh.

Alternatively, it can be replaced by this check:
```suggestion
   if (fes->GetTypicalFE()->GetRangeType() == mfem::FiniteElement::VECTOR)
```